targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/device/drivers/TZ-TRNG/host/src/tztrng_lib/tztrng_driver.c

```diff
@@ -0,0 +1,113 @@
+#include "tztrng_defs.h"
+#include "tztrng_pal.h"
+
+unsigned long gCcRegBase = 0;
+
+static CCError_t RNG_PLAT_SetUserRngParameters(CCRndParams_t *pTrngParams)
+{
+    CCError_t  error = CC_OK;
+
+    /* Set TRNG parameters */
+    pTrngParams->SubSamplingRatio1 = CC_CONFIG_SAMPLE_CNT_ROSC_1;
+    pTrngParams->SubSamplingRatio2 = CC_CONFIG_SAMPLE_CNT_ROSC_2;
+    pTrngParams->SubSamplingRatio3 = CC_CONFIG_SAMPLE_CNT_ROSC_3;
+    pTrngParams->SubSamplingRatio4 = CC_CONFIG_SAMPLE_CNT_ROSC_4;
+
+    /* Allowed ROSCs lengths b'0-3. If bit value 1 - appropriate ROSC is allowed. */
+    pTrngParams->RoscsAllowed = (((pTrngParams->SubSamplingRatio1 > 0) ? 0x1 : 0x0) |
+            ((pTrngParams->SubSamplingRatio2 > 0) ? 0x2 : 0x0) |
+            ((pTrngParams->SubSamplingRatio3 > 0) ? 0x4 : 0x0) |
+            ((pTrngParams->SubSamplingRatio4 > 0) ? 0x8 : 0x0));
+    pTrngParams->SubSamplingRatio = 0;
+
+    if (0 == pTrngParams->RoscsAllowed) {
+        TRNG_LOG_DEBUG("No ROSC allowed!!\n");
+        return LLF_RND_TRNG_REQUIRED_ROSCS_NOT_ALLOWED_ERROR;
+    }
+
+    return error;
+}
+
+uint32_t CC_TrngGetSource(unsigned long rngRegBase, uint8_t *outAddr, size_t *outLen, size_t reqBits)
+{
+    CCError_t Err = CC_OK;
+    uint32_t rndWorkBuff[CC_RND_WORK_BUFFER_SIZE_WORDS];
+    CCRndParams_t rndParams;
+    CCRndState_t rndState = {0};
+    uint32_t  *rndSrc_ptr;
+    uint32_t  sourceOutSizeBytes = 0;
+    int requireBytes = (reqBits % 8) ? (reqBits / 8 + 1) : (reqBits / 8);
+
+    if (NULL == outAddr || NULL == outLen) {
+        TRNG_LOG_DEBUG("NULL pointer!!\n");
+        return LLF_RND_TRNG_ILLEGAL_PTR_ERROR;
+    }
+
+    /* zeroe the rnd buff  */
+    tztrng_memset((uint8_t *)rndWorkBuff, 0, CC_RND_WORK_BUFFER_SIZE_WORDS * sizeof(uint32_t));
+    tztrng_memset((uint8_t *)&rndParams, 0, sizeof(CCRndParams_t));
+    gCcRegBase = rngRegBase;
+
+    if (gCcRegBase == 0) {
+        TRNG_LOG_DEBUG("register base not initialized\n");
+        return LLF_RND_TRNG_ILLEGAL_PTR_ERROR;
+    }
+
+    /* Get parameters and set them in the RND structures */
+    Err = RNG_PLAT_SetUserRngParameters(&rndParams);
+    if (Err != CC_OK)
+        return Err;
+
+    Err = LLF_RND_RunTrngStartupTest(&rndState, &rndParams, rndWorkBuff);
+    if (Err != CC_OK)
+        return Err;
+
+    *outLen = requireBytes;
+    while (requireBytes > 0) {
+        /* Get TRNG Source */
+        Err = LLF_RND_GetTrngSource(
+                &rndState,
+                &rndParams,
+                0,
+                &rndSrc_ptr,
+                &sourceOutSizeBytes,
+                rndWorkBuff);
+
+        if (Err) {
+            TRNG_LOG_DEBUG("LLF_RND_GetTrngSource failed, err[0x%X]\n", (unsigned int)Err);
+            /* memset 0 to outAddr for security concern */
+            tztrng_memset(outAddr, 0, *outLen);
+            *outLen = 0;
+            return Err;
+        }
+
+        /* try to reduce the memcpy */
+        if (requireBytes < (int)sourceOutSizeBytes)
```

Comment:
Strictly speaking, this is a dangerous cast - if `LLF_RND_GetTrngSource` is compromised and sets `sourceOutSizeBytes` to a value above `MAX_INT`, then `(int) sourceOutSizeBytes` may be negative, and the resulting `else` branch performs - which uses `sourceOutSizeBytes` as a `uint32_t` - performs a fauly `memcpy`.